### DISCUSSION TITLE
[General] Clarification Needed: Address Representation as uint256 in zkSync Transactions and Type Conversions

### DISCUSSION THREAD

**Discussion by rholterhus** [zkSync-Community-Hub/zksync-developers#201] [General] [answered]
### Environment

Mainnet

### Provide a brief description of the functionality you're trying to implement.

In the [zkSync `Transaction` struct](https://github.com/matter-labs/era-system-contracts/blob/6ce4f8f44d03bb940c7eee5da63aaabc44682e52/contracts/libraries/TransactionHelper.sol#L25-L71), the `from` and `to` fields are defined as `uint256` values. The [EIP-712 transaction type](https://era.zksync.io/docs/reference/concepts/transactions.html#eip-712-0x71) similarly defines `from` and `to` as `uint256` values in the type structure. I would have expected these fields to be defined as `address` values (equivalent to `uint160` under-the-hood), so I have some questions about handling the necessary type conversions.

Firstly, is it correct that the bootloader will enforce that `from` and `to` are valid EVM addresses (i.e. not larger than `(2 ** 160) - 1`)? I think this is what is happening [here](https://github.com/matter-labs/era-system-contracts/blob/3377d27d7dc26b9f0e1ec0637af34dbc4cb8c2e3/bootloader/bootloader.yul#L3101-L3109), and in this case, it seems that native zkSync transactions can't have weird issues with `to` and `from` values overflowing.

Secondly, is it recommended that other smart contracts using the `Transaction` struct should similarly enforce that `from` and `to` are valid EVM addresses? For example, in [this line](https://github.com/matter-labs/tutorials/blob/b8bf49ca75746ea7e226c34109def89e60387fa6/custom-aa/contracts/TwoUserMultisig.sol#L92) of the tutorial AA wallet, it seems possible that `address(uint160(_transaction.to))` could overflow in the `executeTransactionFromOutside()` path (since the bootloader wouldn't have checked the `Transaction` struct in this case). Would it make sense to change this to a safecast, so that the account can't accept any overflowed addresses?

### What is the specific issue or error you're encountering?

Trying to understand how to custom AA should work with the `from` and `to` address values

### Can you share the error messages or logs you're receiving, if any?

N/A

### Have you made any recent changes to the contract before encountering this issue?

No

### Are there any external libraries or contracts that your contract interacts with?

No

### Can you provide the relevant portions of your contract code where the issue is occurring?

N/A

### Have you tried to isolate the problem, and if so, what were the results?

N/A

### What steps have you already taken to try to resolve the issue?

Tested out the `executeTransactionFromOutside()` path using a `to` address that overflows, and it seems like this is technically allowed

**Comment by MexicanAce** [comment, marked as answer]
Hi @rholterhus 👋 !

> Firstly, is it correct that the bootloader will enforce that `from` and `to` are valid EVM addresses [...] ?

Yes, users should assume that all addresses are valid 160-bit EVM addresses

> Secondly, is it recommended that other smart contracts using the `Transaction` struct should similarly enforce that `from` and `to` are valid EVM addresses?

The decision to have `uint256` was more for forward-compatibility with potential migration to 32-byte addresses https://ethereum-magicians.org/t/increasing-address-size-from-20-to-32-bytes/5485

**Comment by keegansa** [comment]
It's important to ensure that the from and to fields in the zkSync Transaction struct, as well as in any related smart contracts, are properly handled to avoid potential issues with overflowing addresses. While the zkSync bootloader enforces validity for from and to addresses, it's advisable for other contracts interacting with these values to also perform validity checks or use safe type conversions to prevent potential overflow issues. This can help maintain the integrity and security of the system.

**Comment by robbmoser19** [comment]
The zkSync Transaction struct defines from and to fields as uint256 values, which may raise concerns about address overflow. 
It's recommended to ensure that these values are valid EVM addresses to prevent any potential issues.

**Comment by springsj** [comment]
Ensure that from and to fields in zkSync transactions are valid EVM addresses (uint160) to prevent potential overflow issues.
 It's recommended to enforce this check in smart contracts using the Transaction struct to maintain consistency and security.

**Comment by inutpediflinda** [comment]
1. The zkSync bootloader enforces that from and to are valid EVM addresses, preventing overflow issues.
2. It's recommended for other smart contracts to similarly enforce valid EVM addresses for safety. Using safecast instead of direct conversion can prevent overflow issues.

**Comment by rafordtetseyd** [comment]
The zkSync Transaction struct defines the from and to fields as uint256 values, which is different from the standard EVM address type. It is expected that the bootloader will enforce that from and to are valid EVM addresses, which means they cannot be larger than (2 ** 160) - 1. Other smart contracts using the Transaction struct should similarly enforce this requirement for safety. In the AA wallet tutorial, it may make sense to change the address(uint160(_transaction.to)) line to a safecast to prevent overflowed addresses. Overall, it is important to handle the necessary type conversions and ensure that the from and to fields are valid EVM addresses to avoid any issues.

**Comment by eddiedenna** [comment]
The user is questioning why the from and to fields in the zkSync Transaction struct and EIP-712 transaction type are defined as uint256 values instead of address values. They are wondering if the bootloader enforces that from and to are valid EVM addresses and if it's recommended for other smart contracts using the Transaction struct to similarly enforce that from and to are valid EVM addresses. The user is also considering changing a line of code in the tutorial AA wallet to prevent overflowed addresses.
